Mathlib/LinearAlgebra/SpecialLinearGroup.lean

```diff
@@ -0,0 +1,605 @@
+  have : Fintype ι := Fintype.ofFinite ι
+  have : DecidableEq ι := Classical.typeDecidableEq ι
+  rw [Module.finrank_eq_card_basis b, Fintype.card_eq_one_iff] at d1
+  obtain ⟨i, hi⟩ := d1
+  exact ⟨{
+    toFun r := r • (b i)
+    invFun v := b.repr v i
+    left_inv r := by simp
+    right_inv v := b.repr.injective <| by ext j; simp [hi j]
+    map_add' r s := by simp [add_smul]
+    map_smul' r s := by simp [mul_smul] }⟩
+
+theorem exists_unique_eq_smul_id_of_finrank_eq_one (d1 : Module.finrank R V = 1) (u : V →ₗ[R] V) :
+    ∃! c : R, u = c • LinearMap.id := by
+  let e := (exists_linearEquiv_of_finrank_eq_one d1).some
+  set c := e.symm (u (e 1)) with hc
+  suffices u = c • LinearMap.id by
+    use c
+    simp only [this, true_and]
+    intro d hcd
+    rw [LinearMap.ext_iff] at hcd
+    specialize hcd (e 1)
+    simp only [LinearMap.smul_apply, LinearMap.id_coe, id_eq] at hcd
+    have hcd' := LinearEquiv.congr_arg (e := e.symm) hcd
+    simp only [map_smul, LinearEquiv.symm_apply_apply, smul_eq_mul, mul_one] at hcd'
+    rw [hcd']
+  ext x
+  have (x : V) : x = (e.symm x) • (e 1) := by simp [← LinearEquiv.map_smul]
+  rw [this x]
+  simp only [hc, map_smul, LinearMap.smul_apply, LinearMap.id_coe, id_eq]
+  rw [← this]
+
+/-- Endomorphisms of a free module of rank one are homotheties. -/
+noncomputable def LinearEquiv.smul_id_of_rank_one (d1 : Module.finrank R V = 1) :
+    R ≃ₗ[R] (V →ₗ[R] V) where
+  toFun := fun c ↦ c • LinearMap.id
+  map_add' c d := by ext; simp [add_smul]
+  map_smul' c d := by ext; simp [mul_smul]
+  invFun u := (exists_unique_eq_smul_id_of_finrank_eq_one d1 u).choose
+  left_inv c := by
+    simp [← (exists_unique_eq_smul_id_of_finrank_eq_one d1 (c • LinearMap.id)).choose_spec.2 c]
+  right_inv u :=
+    ((exists_unique_eq_smul_id_of_finrank_eq_one d1 u).choose_spec.1).symm
+
+theorem subsingleton_of_subsingleton (d1 : Module.finrank R V = 1) :
+    Subsingleton (SpecialLinearGroup R V) where
+  allEq u v := by
+    ext x
+    by_cases hx : x = 0
+    · simp [hx]
+    suffices ∀ (u : SpecialLinearGroup R V), (u : V →ₗ[R] V) = LinearMap.id by
+      simp only [LinearMap.ext_iff, LinearEquiv.coe_coe, LinearMap.id_coe, id_eq] at this
+      simp
+      rw [this u, this v]
+    intro u
+    ext x
+    set c := (LinearEquiv.smul_id_of_rank_one d1).symm u with hc
+    rw [LinearEquiv.eq_symm_apply] at hc
+    suffices c = 1 by
+      simp [← hc, LinearEquiv.smul_id_of_rank_one, this]
+    have hu := u.prop
+    simpa [← Units.val_inj, LinearEquiv.coe_det, ← hc,
+      LinearEquiv.smul_id_of_rank_one, d1] using hu
+
+end rankOne
+
+instance hasInv : Inv (SpecialLinearGroup R V) :=
+  ⟨fun A => ⟨A⁻¹, by simp [A.prop]⟩⟩
+
+instance hasMul : Mul (SpecialLinearGroup R V) :=
+  ⟨fun A B => ⟨A * B, by simp [A.prop, B.prop]⟩⟩
+
+instance hasDiv : Div (SpecialLinearGroup R V) :=
+  ⟨fun A B => ⟨A / B, by simp [A.prop, B.prop]⟩⟩
+
+instance hasOne : One (SpecialLinearGroup R V) :=
+  ⟨⟨1, by simp⟩⟩
```

Comment:
`has` prefixes are a remnant of lean 3; please use `instFoo` if you need to name them